### PR TITLE
Feature 163666888 update winex functionality

### PR DESCRIPTION
**Pull request by sandinmyjoints**
@ahaurw01 

- pn-logging's winex.js in master comes from winex v0.0.6
- neodarwin's winex.coffee came from winex v0.0.6, too, but it gained some functionality over the years
- this PR updates pn-logging's winex.js to match the functionality that neodarwin's winex.coffee has, as preparation for using pn-logging in neodarwin
- to do this, I ran this in neodarwin: `coffee -b -c -o ../pn-logging ./src/server/winex.coffee`

Here's a diff of winex v0.0.6 winex.coffee vs. neodarwin's winex.coffee:

```diff
--- ./winex.coffee	2019-01-09 18:37:52.000000000 -0500
+++ ../neodarwin/src/server/winex.coffee	2018-11-19 15:27:36.000000000 -0500
@@ -111,7 +111,7 @@
 #
 # @param        [Object] winstonLogger Winston logger.
 # @param        [Object] classMeta     Extra meta for every log call.
-# @param        [Object] opts          Extra meta for every log call.
+# @param        [Object] opts          Extra options for every log call.
 # @config opts  [Object] nop           Use NOP logger instead.
 factory = (winstonLogger, classMeta = {}, opts = {}) ->
   useNop = opts.nop is true
@@ -125,9 +125,11 @@
       @meta       = extend {}, (opts.meta ? {})
       @type       = opts.type ? null
       @errNoStack = opts.errNoStack is true
+      @info404    = opts.info404 is true
 
       # Patch in incoming data.
       @addReq   opts.req    if opts.req
+      @addRes   opts.res    if opts.res
       @addError opts.error  if opts.error
 
     # Default request type if not specified.
@@ -155,10 +157,13 @@
 
           level = "info"
           if 400 <= res.statusCode < 500
-            level = "warning"
+            level = "warn" unless (res.statusCode is 404 and log.info404)
           if res.statusCode >= 500
             level = "error"
 
+          # Allow overriding the level.
+          level = log.level if log.level?
+
           log.addRes res
           log[level] "request"
 
@@ -168,12 +173,14 @@
     addReq: (req = {}) ->
       maxChars    = 200
       urlObj      = url.parse req?.url
-      path        = urlObj?.pathname
+      path        = urlObj?.pathname ? ""
       query       = urlObj?.query ? ""
-      queryChars  = query?.length
-      pathLength  = path?.length ? 0
-      path        = path.substr  0, maxChars if pathLength > maxChars
-      query       = query.substr 0, maxChars if queryChars > maxChars
+      queryChars  = Array.from(query).length
+      pathLength  = Array.from(path).length
+      if pathLength > maxChars
+        path  = Array.from(path).slice(0, maxChars).join('')
+      if queryChars > maxChars
+        query = Array.from(query).slice(0, maxChars).join('')
 
       @meta = extend @meta, stripNulls({
         reqClient:      clientIp req
@@ -198,9 +205,11 @@
 
       unless @errNoStack
         extend errObj, stripNulls({
-          errArgs:  (error?.arguments ? "").toString().substr(0, 100)
+          errArgs:  Array.from((error?.arguments ? "").toString()).slice(
+            0, 100).join('')
           errType:  error?.type ? null
           errStack: error?.stack ? null
+          errCode:  error?.code ? null
           errKnown: 0
         })
 
@@ -213,7 +222,7 @@
 
     # Modify and aggregate list of metas.
     #
-    @patchMeta: (metas...) =>
+    @patchMeta: (metas...) ->
       # Filter the metas
       metas = (m for m in metas when m?)
 
```

The differences, and what this PR does about them:
- [X] Neodarwin has info404 -- adds it
- [X] Neodarwin uses warn not warning -- keeps warning. Neodarwin will switch to warning when it starts using pn-logging.
- [X] Neodarwin has addRes -- adds it
- [X] Neodarwin allows overriding the level with log.level if set -- adds it
- [X] Neodarwin defaults path to "" if pathname is null -- adds it
- [X] Neodarwin uses Array.from so length fields are multibyte aware -- adds it
- [X] Neodarwin addError looks for error.code -- adds it

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -2,6 +2,9 @@
   "name": "pn-logging",
   "version": "5.0.0",
   "description": "A logging wrapper around winston and sentry.",
+  "engines": {
```
Comment (sandinmyjoints) [inline package.json:5]: Due to `Array.from`.

File: test/index.js [hunk 2]
```diff
@@ -158,11 +229,14 @@ describe('index', function() {
     });
 
     function makeReq(props) {
-      return Object.assign({url: 'http://example.com/path', path: '/path'}, props)
+      return Object.assign({ url: 'http://example.com/path' }, props);
```
Comment (sandinmyjoints) [inline test/index.js:230]: From here down, in this file, some of the changes are just due to prettier. I'll mark where there's new stuff.

File: test/index.js [hunk 3]
```diff
@@ -172,74 +246,193 @@ describe('index', function() {
...
+          reqPath: '/path',
+          reqQuery: '',
+          reqQueryChars: 0,
+          resStatus: '400',
+        });
+        done();
+      });
+    });
+
+    it('should patch res.end and emit a warning log instance for a 404 response', function(done) {
```
Comment (sandinmyjoints) [inline test/index.js:295]: New test.

File: test/index.js [hunk 4]
```diff
@@ -172,74 +246,193 @@ describe('index', function() {
...
+          reqPath: '/path',
+          reqQuery: '',
+          reqQueryChars: 0,
+          resStatus: '404',
+        });
+        done();
+      });
+    });
+
+    it('should patch res.end and emit an info log instance for a 404 response with info404 option', function(done) {
```
Comment (sandinmyjoints) [inline test/index.js:322]: New test.

File: test/index.js [hunk 5]
```diff
@@ -172,74 +246,193 @@ describe('index', function() {
...
+          reqPath: '/path',
+          reqQuery: '',
+          reqQueryChars: 0,
+          resStatus: '500',
+        });
+        done();
+      });
+    });
+
+    it('should allow overriding the level with log.level', function(done) {
```
Comment (sandinmyjoints) [inline test/index.js:377]: New test.

File: test/index.js [hunk 6]
```diff
@@ -172,74 +246,193 @@ describe('index', function() {
...
+          reqPath: '/path',
+          reqQuery: '',
+          reqQueryChars: 0,
+          resStatus: '200',
+        });
+        done();
+      });
+    });
+
+    it('should correctly count multibyte characters in url query', function(done) {
```
Comment (sandinmyjoints) [inline test/index.js:406]: New test.

File: test/index.js [hunk 7]
```diff
@@ -141,6 +141,75 @@ describe('index', function() {
     });
   });
 
+  describe('Winex Log class', () => {
```
Comment (sandinmyjoints) [inline test/index.js:144]: The test suite that winex never had.

**Review by ahaurw01 — APPROVED** [review]
Thanks for adding some tests! I integrated this version with sd-playground as a sanity check. Looks good.

**Comment by sandinmyjoints** [comment]
Thanks, that was a good idea!

**Pull request merged**